File: gisaid_track_aa_subs_2.jl
#=
gisaid_track_aa_subs_2.jl
    Create a CSV file of AA substitutions in human sequences by date and country from a
    GISAID metafile.
    version 1.1

@author Bill Thompson
@copyright 2021_04-28
@dlicense GPL 3
=#

using DataFrames
using CSV
using Dates
using ArgParse

function GetArgs()
    # read arguments from command line
    # from https://argparsejl.readthedocs.io/en/latest/argparse.html
    s = ArgParseSettings(description = "Extract AA substituions from GISAID metefile.",
                         version = "1.1",
                         add_version = true)

    @add_arg_table s begin
        "--input_file", "-i"
        required = true
        help = "GISAID meteafile (required)."
        arg_type = String

        "--output_file", "-o"
        required = true
        help = "CSV file with results (required)."
        arg_type = String

        "--host", "-s"
        required = false
        help = "Host species (default = Human)."
        arg_type = String
        default = "Human"
    end

    return parse_args(s)
end

function check_date(date_str)
    #=
    Check that data is formatted as yyyy-mm-dd.

    @arguments
    data_str - String
        A string representing a date.

    @returns
    A Date in yyyy-mm-dd of nothing if string is not properly formatted.
    =#
    r = r"^\d\d\d\d\-\d\d\-\d\d$"
    if isnothing(match(r, date_str))
        return nothing
    end

    return Date(date_str, "yyyy-mm-dd")
end

function country_from_location(location_str)
    #=
    Split and trim country from a GISAID Location. 
    Country is the second term in Location.

    @arguments
    location_str - String   
        A Location value from GISAID.
        
    @returns
    A String containing the country name.
    =#
    return strip(split(location_str, "/")[2], ' ')
end

function main()
    #=
    Check sequences for valid date and host. 
    Write a CSV file with a row for each AA substitution.
    =#
    args = GetArgs()
    meta_table = args["input_file"]
    aa_subs_table = args["output_file"]
    host = args["host"]
   
    # Read the metfile.
    df = DataFrame(CSV.File(meta_table, delim = '\t'))
    
    # select columns and filter the data
    df2 = df[!, ["Virus name", "Accession ID", 
                "Collection date", "Location",
                "Pango lineage", "AA Substitutions",
                "Host"]]

    
    df2 = df2[df2[!, :Host] .== host, :] # filter host and remove NAs from lineade
    df2 = dropmissing(df2, "Pango lineage")

    # Throw out invalid dates
    df2[!, "Date"] = check_date.(df2[!, "Collection date"])
    df2 = df2[df2[!, "Date"] .!== nothing, :]

    df2[!, "Country"] = country_from_location.(df2[!, "Location"])

    # Clean up AA substitutions and explode AA into rows
    df2[!, "AA Substitutions"] = chop.(df2[!, "AA Substitutions"], head = 1, tail = 1)
    df2[!, "AA Substitutions"] = split.(df2[!, "AA Substitutions"], ",")
    df2 = flatten(df2, ["AA Substitutions"])

    select!(df2, Not(["Location", "Host", "Collection date"])) # We don't need these

    CSV.write(aa_subs_table, df2) 
end

main()

File: gisaid_track_aa_subs_2.py
#!/home/bill/anaconda3/envs/bio2/bin/python
# -*- coding: utf-8 -*-
"""
gisaid_track_aa_subs.py
    Create a CSV file of AA substitutions in human sequences by date and country from a
    GISAID metafile.

@author: Bill Thompson
@license: GPL 3
@copyright: 2021_04_28
"""
import argparse
import sys
import re
import datetime
import pandas as pd

def GetArgs():
    def ParseArgs(parser):
        class Parser(argparse.ArgumentParser):
            def error(self, message):
                sys.stderr.write('error: %s\n' % message)
                self.print_help()
                sys.exit(2)
                
        parser = Parser(description='Reformat GISAID metefile.')
        parser.add_argument('-i', '--input_file',
                            required = True,
                            help = 'GISAID metafile (required).',
                            type = str)
        parser.add_argument('-o', '--output_file',
                            required = True,
                            help = 'Output CSV file.',
                            type = str)
        parser.add_argument('-s', '--host',
                            required = False,
                            help = 'Virus host species (default = Human).',
                            default = "Human",
                            type = str)

        return parser.parse_args()

    parser = argparse.ArgumentParser()
    args = ParseArgs(parser)
    
    return args

def check_date(date_str):
    """
    Check that data is formatted as yyyy-mm-dd.

    @arguments
    data_str - String
        A string representing a date.

    @returns
    A Date in yyyy-mm-dd of nothing if string is not properly formatted.
    """
    p = re.compile(r"^\d\d\d\d\-\d\d\-\d\d$")
    if p.match(date_str) is None:
        return None

    return datetime.datetime.strptime(date_str, "%Y-%m-%d")

def main():
    args = GetArgs()
    meta_file = args.input_file
    out_file = args.output_file
    host = args.host
    
    df = pd.read_csv(meta_file,
                     sep = '\t', 
                     dtype={'Location': str, "Variant": str, "Is reference?": str})

    # create an empty dataframe
    df2 = pd.DataFrame(columns = ["Date", "Country", "Virus_name",
                                  "Accession_ID", "Pango_lineage",
                                  "AA_subs"])

    for i in range(df.shape[0]):
        if i % 10000 == 0:
            print(i)
        
        if df.loc[i, "Host"] != host:
            continue

        if df.loc[i, "Pango lineage"] == "":
            continue

        date = check_date(df.loc[i, "Collection date"])
        if date is None:
            continue

        country = df.loc[i, "Location"].split("/")[1].strip()
        
        aas = df.loc[i, "AA Substitutions"][1:len(df.loc[i, "AA Substitutions"])-1].split(",")
        if aas is None:
            continue

        for aa in aas:
            df2 = df2.append({"Date": date,
                              "Country": country, 
                              "Virus_name": df.loc[i, "Virus name"],
                              "Accession_ID": df.loc[i, "Accession ID"],
                              "Pango_lineage": df.loc[i, "Pango lineage"],
                              "AA_sub": aa}, ignore_index = True)
    
    df2.to_csv(out_file, index = False)

if __name__ == '__main__':
    main()